saleor/graphql/discount/mutations/sale/sale_channel_listing_update.py

```diff
@@ -73,18 +82,75 @@ class Meta:
         error_type_field = "discount_errors"
 
     @classmethod
-    def add_channels(cls, sale: "SaleModel", add_channels: List[Dict]):
+    def add_channels(cls, promotion: Promotion, add_channels: List[Dict]):
+        rules_data_to_add: List[RuleInfo] = []
+        rules_to_update: List[PromotionRule] = []
+
+        rules = promotion.rules.all()
+        PromotionRuleChannel = PromotionRule.channels.through
+        rule_ids = [rule.id for rule in rules]
+        rule_channel = PromotionRuleChannel.objects.filter(
+            promotionrule_id__in=rule_ids
+        )
+        current_channel_ids = rule_channel.values_list("channel_id", flat=True)
+        examplart_rule = rules[0]
         for add_channel in add_channels:
             channel = add_channel["channel"]
-            defaults = {"currency": channel.currency_code}
-            channel = add_channel["channel"]
-            if "discount_value" in add_channel.keys():
-                defaults["discount_value"] = add_channel.get("discount_value")
-            SaleChannelListing.objects.update_or_create(
-                sale=sale,
-                channel=channel,
-                defaults=defaults,
+            discount_value = add_channel["discount_value"]
+
+            if channel.id not in current_channel_ids:
+                rules_data_to_add.append(
+                    RuleInfo(
+                        rule=PromotionRule(
+                            name=examplart_rule.name,
+                            promotion=promotion,
+                            catalogue_predicate=examplart_rule.catalogue_predicate,
+                            reward_value_type=examplart_rule.reward_value_type,
+                            reward_value=discount_value,
+                        ),
+                        channel=channel,
+                    )
+                )
+            else:
+                # We ensure that every rule has one or none related channel
+                rule_to_update = rules.get(
+                    id=rule_channel.get(channel_id=channel.id).promotionrule_id  # type: ignore[attr-defined] # noqa: E501
+                )
```

Comment:
IMO it's hard to understand. I'll suggest preparing some mapping before like:
```
        channel_to_rule_mapping = {
            channel_id: rule_id for channel_id, rule_id in rule_channel.values_list("channel_id", "promotionrule_id")
        }
```

Then `current_channel_ids` could be replaces with `channel_to_rule_mapping.keys()`